controllers/appgroup_controller.go

```diff
@@ -505,26 +525,63 @@ func (r *ApplicationGroupReconciler) cleanupWorkflow(ctx context.Context, logr l
 		for _, node := range wf.Status.Nodes {
 			nodes[node.ID] = node
 		}
-		graph, err := workflow.Build(g.Name, nodes)
+		rwf := &v1alpha12.Workflow{}
+
+		rwfName := fmt.Sprintf("%s-reverse", wf.Name)
+		rwfNamespace := wf.Namespace
+		err := r.Client.Get(ctx, types.NamespacedName{Namespace: rwfNamespace, Name: rwfName}, rwf)
 		if err != nil {
-			logr.Error(err, "failed to build the wf status DAG")
-		}
+			if kerrors.IsNotFound(err) {
+				err = r.generateReverseWorkflow(ctx, logr, nodes, &g)
+				if err != nil {
+					logr.Error(err, "failed to generate reverse workflow")
+					return err
+				}
 
-		rev := graph.Reverse()
+				// update the forward workflow metadata with a finalizer
+				wf.Finalizers = []string{rwffinalizer}
 
-		for _, bucket := range rev {
-			for _, hr := range bucket {
-				err = r.Client.Delete(ctx, &hr)
+				wfPatch := client.MergeFrom(wf.DeepCopy())
```

Comment:
I thought this was needed after adding (or removing) the finalizer here so that the change is reflected on the object on the cluster. @nitishm - did I take your feedback incorrectly?

Reply:
Yes, my understanding was that you would need to do a `DeepCopy` before you update the object, otherwise the mergedFrom object and the updated object will not be any different